numba/core/interpreter.py

```diff
@@ -80,6 +78,229 @@ def get_assignment_source(self, destname):
+                    for x in peep_hole:
+                        print(x)
+                    print("")
+
+                appends = []
+                extends = []
+                init = region[1][1]
+                const_list = init.target.name
+                # Walk through the peep_hole and find things that are being
+                # "extend"ed and "append"ed to the BUILD_LIST
+                for x in peep_hole:
+                    if isinstance(x, ir.Assign):
+                        if isinstance(x.value, ir.Expr):
+                            expr = x.value
+                            if (expr.op == 'getattr' and
+                                    expr.value.name == const_list):
+                                # it's not strictly necessary to split out
+                                # extends and appends, but it helps with
+                                # debugging to do so!
+                                if expr.attr == 'extend':
+                                    extends.append(x.target.name)
+                                elif expr.attr == 'append':
+                                    appends.append(x.target.name)
+                                else:
+                                    assert 0
+                # go back through the peep hole build new IR based on it.
+                new_hole = []
+
+                def append_and_fix(x):
+                    """ Adds to the new_hole and fixes up definitions"""
+                    new_hole.append(x)
+                    if x.target.name in func_ir._definitions:
+                        # if there's already a definition, drop it, should only
+                        # be 1 as the way cpython emits the sequence for
+                        # `list_to_tuple` should ensure this.
+                        assert len(func_ir._definitions[x.target.name]) == 1
+                        func_ir._definitions[x.target.name].clear()
+                    func_ir._definitions[x.target.name].append(x.value)
+
+                the_build_list = init.target
+
+                # Do the transform on the peep hole
+                if _DEBUG:
+                    print("\nBLOCK:")
+                    blk.dump()
+
+                # This section basically accumulates list appends and extends
+                # as binop(+) on tuples, it drops all the getattr() for extend
+                # and append as they are now dead and replaced with binop(+).
+                # It also switches out the build_list for a build_tuple and then
+                # ensures everything is wired up and defined ok.
+                t2l_agn = region[0][1]
+                acc = the_build_list
+                for x in peep_hole:
+                    if isinstance(x, ir.Assign):
+                        if isinstance(x.value, ir.Expr):
+                            expr = x.value
+                            if expr.op == 'getattr':
+                                if expr.attr in ('extend', 'append'):
```

Comment:
This didn't check that the expr is related to the the const list. My suggestion: https://github.com/sklam/numba/commit/960bc6783881d7971b1a1362d3ecfe6dc50a1a50

Reply:
This also made me think about a case like:
```python
@njit
def foo(a):
    z = [1, 2]
    return (*a, z.append(3), z.extend(a), np.ones(3)), z

x = (4, 5.6)
foo(x)
```
In this case there's other calls and append/extend in the unpack that do not relate to the unpacked tuple target . I'm not sure how common this sort of thing is but it's legal so I've patched up to make it work.